src/mme-app/actionHandlers/s1HandoverActionHandlers.cpp

```diff
@@ -0,0 +1,659 @@
+    if (msgData_p == NULL)
+    {
+        log_msg(LOG_ERROR, "Failed to retrieve data buffer \n");
+        return ActStatus::HALT;
+    }
+
+    const struct handover_notify_Q_msg &ho_notify =
+            (msgData_p->msg_data.handover_notify_Q_msg_m);
+
+    // The UE has synced to target cell. Set the current enb
+    // to target enb.
+    ue_ctxt->setEnbFd(ho_ctxt->getTargetEnbContextId());
+    ue_ctxt->setS1apEnbUeId(ho_ctxt->getTargetS1apEnbUeId());
+
+    // Wait till TAU complete to overwrite TAI?
+    //TAI and CGI obtained from s1ap ies.
+    //Convert the PLMN in s1ap format to nas format before storing in procedure context.
+    MmeCommonUtils::formatS1apPlmnId(const_cast<PLMN*>(&ho_notify.tai.plmn_id));
+    MmeCommonUtils::formatS1apPlmnId(const_cast<PLMN*>(&ho_notify.utran_cgi.plmn_id));
+    ho_ctxt->setTargetTai(Tai(ho_notify.tai));
+    ho_ctxt->setTargetCgi(Cgi(ho_notify.utran_cgi));
+
+    ProcedureStats::num_of_ho_notify_received++;
+    log_msg(LOG_DEBUG, "Leaving process_ho_notify\n");
+
+    return ActStatus::PROCEED;
+}
+
+/********************************************
+ * Action handler : send_mb_req_to_sgw_for_ho
+ ********************************************/
+ActStatus ActionHandlers::send_mb_req_to_sgw_for_ho(ControlBlock &cb)
+{
+    log_msg(LOG_DEBUG, "Inside send_mb_req_to_sgw_for_ho \n");
+
+    UEContext *ue_ctxt = static_cast<UEContext*>(cb.getPermDataBlock());
+    if (ue_ctxt == NULL)
+    {
+        log_msg(LOG_DEBUG,
+                "send_mb_req_to_sgw_for_ho: ue context or procedure ctxt is NULL \n");
+        return ActStatus::HALT;
+    }
+
+    S1HandoverProcedureContext *ho_ctxt =
+            dynamic_cast<S1HandoverProcedureContext*>(cb.getTempDataBlock());
+    if (ho_ctxt == NULL)
+    {
+        log_msg(LOG_DEBUG, "process_ho_notify: procedure ctxt is NULL \n");
+        return ActStatus::HALT;
+    }
+    SessionContext *sessionCtxt = ue_ctxt->getSessionContext();
+    if (sessionCtxt == NULL)
+    {
+        log_msg(LOG_DEBUG,
+                "send_mb_req_to_sgw_for_ho: session ctxt is NULL \n");
+        return ActStatus::HALT;
+    }
+
+    struct MB_Q_msg mb_msg;
+    MmeGtpMsgUtils::populateModifyBearerRequestHo(
+            cb, *ue_ctxt, *sessionCtxt, *ho_ctxt, mb_msg);
+
+    cmn::ipc::IpcAddress destAddr = {TipcServiceInstance::s11AppInstanceNum_c};
+    MmeIpcInterface &mmeIpcIf = static_cast<MmeIpcInterface&>(compDb.getComponent(MmeIpcInterfaceCompId));
+    mmeIpcIf.dispatchIpcMsg((char *) &mb_msg, sizeof(mb_msg), destAddr);
+
+    ProcedureStats::num_of_mb_req_to_sgw_sent++;
+    log_msg(LOG_DEBUG, "Leaving send_mb_req_to_sgw_for_ho \n");
+    return ActStatus::PROCEED;
+}
+
+/***************************************
+* Action handler : process_mb_resp_for_ho
+***************************************/
+ActStatus ActionHandlers::process_mb_resp_for_ho(ControlBlock& cb)
+{
+    log_msg(LOG_DEBUG, "Inside process_mb_resp_for_ho \n");
```

Comment:
HO failure can happen because negative response from SGW...Please open a defect and we should handle it at some time.